Fix #109: synthesize mappings from #error directives

kimgr opened the pull request:
cc @sylvestre 

Like with `@headername`, we now parse header files to find known `#error`
directives for GNU libc and the glib library.

These `#error` directives are phrased in a way that we can construct
mappings from private to public headers automatically, instead of
maintaining manual mappings.

I've tested this for `glib`, which I happened to have installed on my FreeBSD rig, but I don't have an environment with `glibc`, so it would be nice to revert the recent `socket_types.h` mapping and re-run the repro case from #245.

sylvestre commented:
On a debian testing (stretch), I am getting a failure:
```
======================================================================
FAIL: runTest (__main__.hash_error)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "run_iwyu_tests.py", line 172, in <lambda>
    {'runTest': lambda self, f=filename: self.RunOneTest(f),
  File "run_iwyu_tests.py", line 143, in RunOneTest
    iwyu_flags, clang_flags, verbose=True)
  File "/home/sylvestre/dev/debian/pkg-llvm/iwyu/iwyu.git/iwyu_test_util.py", line 437, in TestIwyuOnRelativeFile
    test_case.assertTrue(not failures, ''.join(failures))
AssertionError: 
tests/cxx/hash_error.cc:26: Unexpected diagnostic:
glib_symbol is defined in "tests/cxx/hash_error-glib-private.h", which isn't directly #included.

tests/cxx/hash_error.cc:26: Unexpected diagnostic:
glibc1_symbol is defined in "tests/cxx/hash_error-glibc1-private.h", which isn't directly #included.

tests/cxx/hash_error.cc:26: Unexpected diagnostic:
glibc2_symbol is defined in "tests/cxx/hash_error-glibc2-private.h", which isn't directly #included.

Unexpected summary diffs for tests/cxx/hash_error.cc:
+++ 
@@ -1,9 +1,15 @@
 tests/cxx/hash_error.cc should add these lines:
+#include "tests/cxx/hash_error-glib-private.h"
+#include "tests/cxx/hash_error-glibc1-private.h"
+#include "tests/cxx/hash_error-glibc2-private.h"
 
 tests/cxx/hash_error.cc should remove these lines:
+- #include <tests/cxx/hash_error-glib-public.h>  // lines XX-XX
+- #include <tests/cxx/hash_error-glibc1-public.h>  // lines XX-XX
+- #include <tests/cxx/hash_error-glibc2-public.h>  // lines XX-XX
 - #include "tests/cxx/hash_error-nomap.h"  // lines XX-XX
 
 The full include-list for tests/cxx/hash_error.cc:
-#include <tests/cxx/hash_error-glib-public.h>  // for glib_symbol
-#include <tests/cxx/hash_error-glibc1-public.h>  // for glibc1_symbol
-#include <tests/cxx/hash_error-glibc2-public.h>  // for glibc2_symbol
+#include "tests/cxx/hash_error-glib-private.h"  // for glib_symbol
+#include "tests/cxx/hash_error-glibc1-private.h"  // for glibc1_symbol
+#include "tests/cxx/hash_error-glibc2-private.h"  // for glibc2_symbol
---

```

kimgr commented:
Thanks for testing! Can you try:

    $ include-what-you-use -Xiwyu -v5 -I . tests/cxx/hash_error.cc

and look at the very beginning of the output, it spits out which `#error` mappings it finds.

It looks suspiciously like the output I got before this patch, are you sure you rebuilt IWYU?

sylvestre commented:
I did rebuilt it (I just checked). My git status: http://paste.debian.net/359485/
The verbose mode:
 ```
$ include-what-you-use -Xiwyu -v5 -I . tests/cxx/hash_error.cc
Setting verbose-level to 5
Added to main compilation unit: tests/cxx/hash_error.cc
[ Exiting to  ] tests/cxx/hash_error.cc:1:1
[ #include    ] <tests/cxx/hash_error-glibc1-public.h> (tests/cxx/hash_error-glibc1-public.h)
Added an #include: tests/cxx/hash_error.cc -> <tests/cxx/hash_error-glibc1-public.h>
[ Exiting to  ] tests/cxx/hash_error.cc:22:1
[ #include    ] <tests/cxx/hash_error-glibc2-public.h> (tests/cxx/hash_error-glibc2-public.h)
Added an #include: tests/cxx/hash_error.cc -> <tests/cxx/hash_error-glibc2-public.h>
[ Exiting to  ] tests/cxx/hash_error.cc:23:1
[ #include    ] <tests/cxx/hash_error-glib-public.h> (tests/cxx/hash_error-glib-public.h)
Added an #include: tests/cxx/hash_error.cc -> <tests/cxx/hash_error-glib-public.h>
[ Exiting to  ] tests/cxx/hash_error.cc:24:1
[ #include    ] "tests/cxx/hash_error-nomap.h" (tests/cxx/hash_error-nomap.h)
Added an #include: tests/cxx/hash_error.cc -> "tests/cxx/hash_error-nomap.h"
[ Exiting to  ] tests/cxx/hash_error.cc:25:1
Intends-to-provide for tests/cxx/hash_error.cc:
   tests/cxx/hash_error.cc
   tests/cxx/hash_error-glibc1-public.h
   tests/cxx/hash_error-glibc2-public.h
   tests/cxx/hash_error-glib-public.h
   tests/cxx/hash_error-nomap.h
Intends-to-provide for tests/cxx/hash_error-glibc1-public.h:
   tests/cxx/hash_error-glibc1-public.h
   tests/cxx/hash_error-glibc1-private.h
Intends-to-provide for tests/cxx/hash_error-glibc1-private.h:
   tests/cxx/hash_error-glibc1-private.h
Intends-to-provide for tests/cxx/hash_error-glibc2-public.h:
   tests/cxx/hash_error-glibc2-public.h
   tests/cxx/hash_error-glibc2-private.h
Intends-to-provide for tests/cxx/hash_error-glibc2-private.h:
   tests/cxx/hash_error-glibc2-private.h
Intends-to-provide for tests/cxx/hash_error-glib-public.h:
   tests/cxx/hash_error-glib-public.h
   tests/cxx/hash_error-glib-private.h
Intends-to-provide for tests/cxx/hash_error-glib-private.h:
   tests/cxx/hash_error-glib-private.h
Intends-to-provide for tests/cxx/hash_error-nomap.h:
   tests/cxx/hash_error-nomap.h
tests/cxx/hash_error.cc:26:12: (1) [ VarDecl ] static int total = glibc1_symbol + glibc2_symbol + glib_symbol
tests/cxx/hash_error.cc:26:8: (2) [ BuiltinType ] int
tests/cxx/hash_error.cc:26:8: (2) [ BuiltinTypeLoc ] int
tests/cxx/hash_error.cc:26:50: (2) [ BinaryOperator ] BinaryOperator 0x3c39060 <tests/cxx/hash_error.cc:26:20, col:52> 'int' '+'
|-BinaryOperator 0x3bc92f8 <col:20, col:36> 'int' '+'
| |-ImplicitCastExpr 0x3bc92c8 <col:20> 'int' <LValueToRValue>
| | `-DeclRefExpr 0x3bc9278 <col:20> 'int' lvalue Var 0x3bc9018 'glibc1_symbol' 'int'
| `-ImplicitCastExpr 0x3bc92e0 <col:36> 'int' <LValueToRValue>
|   `-DeclRefExpr 0x3bc92a0 <col:36> 'int' lvalue Var 0x3bc90c8 'glibc2_symbol' 'int'
`-ImplicitCastExpr 0x3bc9348 <col:52> 'int' <LValueToRValue>
  `-DeclRefExpr 0x3bc9320 <col:52> 'int' lvalue Var 0x3bc9138 'glib_symbol' 'int'

tests/cxx/hash_error.cc:26:50: (2) [ BinaryOperator ] BinaryOperator 0x3bc92f8 <tests/cxx/hash_error.cc:26:20, col:36> 'int' '+'
|-ImplicitCastExpr 0x3bc92c8 <col:20> 'int' <LValueToRValue>
| `-DeclRefExpr 0x3bc9278 <col:20> 'int' lvalue Var 0x3bc9018 'glibc1_symbol' 'int'
`-ImplicitCastExpr 0x3bc92e0 <col:36> 'int' <LValueToRValue>
  `-DeclRefExpr 0x3bc92a0 <col:36> 'int' lvalue Var 0x3bc90c8 'glibc2_symbol' 'int'

tests/cxx/hash_error.cc:26:20: (3) [ ImplicitCastExpr ] ImplicitCastExpr 0x3bc92c8 <tests/cxx/hash_error.cc:26:20> 'int' <LValueToRValue>
`-DeclRefExpr 0x3bc9278 <col:20> 'int' lvalue Var 0x3bc9018 'glibc1_symbol' 'int'

tests/cxx/hash_error.cc:26:20: (4) [ DeclRefExpr ] DeclRefExpr 0x3bc9278 <tests/cxx/hash_error.cc:26:20> 'int' lvalue Var 0x3bc9018 'glibc1_symbol' 'int'

tests/cxx/hash_error.cc:26:36: (3) [ ImplicitCastExpr ] ImplicitCastExpr 0x3bc92e0 <tests/cxx/hash_error.cc:26:36> 'int' <LValueToRValue>
`-DeclRefExpr 0x3bc92a0 <col:36> 'int' lvalue Var 0x3bc90c8 'glibc2_symbol' 'int'

tests/cxx/hash_error.cc:26:36: (4) [ DeclRefExpr ] DeclRefExpr 0x3bc92a0 <tests/cxx/hash_error.cc:26:36> 'int' lvalue Var 0x3bc90c8 'glibc2_symbol' 'int'

tests/cxx/hash_error.cc:26:52: (3) [ ImplicitCastExpr ] ImplicitCastExpr 0x3bc9348 <tests/cxx/hash_error.cc:26:52> 'int' <LValueToRValue>
`-DeclRefExpr 0x3bc9320 <col:52> 'int' lvalue Var 0x3bc9138 'glib_symbol' 'int'

tests/cxx/hash_error.cc:26:52: (4) [ DeclRefExpr ] DeclRefExpr 0x3bc9320 <tests/cxx/hash_error.cc:26:52> 'int' lvalue Var 0x3bc9138 'glib_symbol' 'int'

tests/cxx/hash_error.cc:26:20: warning: glibc1_symbol is defined in "tests/cxx/hash_error-glibc1-private.h", which isn't directly #included.
tests/cxx/hash_error.cc:26:36: warning: glibc2_symbol is defined in "tests/cxx/hash_error-glibc2-private.h", which isn't directly #included.
tests/cxx/hash_error.cc:26:52: warning: glib_symbol is defined in "tests/cxx/hash_error-glib-private.h", which isn't directly #included.

tests/cxx/hash_error.cc should add these lines:
#include "tests/cxx/hash_error-glib-private.h"    // for glib_symbol
#include "tests/cxx/hash_error-glibc1-private.h"  // for glibc1_symbol
#include "tests/cxx/hash_error-glibc2-private.h"  // for glibc2_symbol

tests/cxx/hash_error.cc should remove these lines:
- #include <tests/cxx/hash_error-glib-public.h>  // lines 23-23
- #include <tests/cxx/hash_error-glibc1-public.h>  // lines 21-21
- #include <tests/cxx/hash_error-glibc2-public.h>  // lines 22-22
- #include "tests/cxx/hash_error-nomap.h"  // lines 24-24

The full include-list for tests/cxx/hash_error.cc:
#include "tests/cxx/hash_error-glib-private.h"    // for glib_symbol
#include "tests/cxx/hash_error-glibc1-private.h"  // for glibc1_symbol
#include "tests/cxx/hash_error-glibc2-private.h"  // for glibc2_symbol
---
`̀``

kimgr commented:
Weird, I don't see the expected log output. And `which include-what-you-use` shows the just-built one, right?

I was expecting output like:

```
$ include-what-you-use -Xiwyu -v5 -I . tests/cxx/hash_error.cc
Setting verbose-level to 5
Added to main compilation unit: tests/cxx/hash_error.cc
[ Exiting to  ] tests/cxx/hash_error.cc:1:1
[ #include    ] <tests/cxx/hash_error-glibc1-public.h> (tests/cxx/hash_error-glibc1-public.h)
Added an #include: tests/cxx/hash_error.cc -> <tests/cxx/hash_error-glibc1-public.h>
Adding #error mapping: "tests/cxx/hash_error-glibc1-private.h" -> <tests/cxx/hash_error-glibc1-public.h>
[ Exiting to  ] tests/cxx/hash_error.cc:22:1
[ #include    ] <tests/cxx/hash_error-glibc2-public.h> (tests/cxx/hash_error-glibc2-public.h)
Added an #include: tests/cxx/hash_error.cc -> <tests/cxx/hash_error-glibc2-public.h>
Adding #error mapping: "tests/cxx/hash_error-glibc2-private.h" -> <tests/cxx/hash_error-glibc2-public.h>
[ Exiting to  ] tests/cxx/hash_error.cc:23:1
[ #include    ] <tests/cxx/hash_error-glib-public.h> (tests/cxx/hash_error-glib-public.h)
Added an #include: tests/cxx/hash_error.cc -> <tests/cxx/hash_error-glib-public.h>
Adding #error mapping: "tests/cxx/hash_error-glib-private.h" -> <tests/cxx/hash_error-glib-public.h>
[ Exiting to  ] tests/cxx/hash_error.cc:24:1
[ #include    ] "tests/cxx/hash_error-nomap.h" (tests/cxx/hash_error-nomap.h)
Added an #include: tests/cxx/hash_error.cc -> "tests/cxx/hash_error-nomap.h"
[ Exiting to  ] tests/cxx/hash_error.cc:25:1
```
Not sure why this would act differently on Debian compared to FreeBSD, seems like pretty basic text-processing.

sylvestre commented:
Oh... Shame on me... The testsuite was not using my fresh iwyu build but the one from my system (from the Debian package)... (probably a bug btw).

It seems to be ok (I have 8 failures now but it is the same number without the patch). Sorry about the noise...

kimgr commented:
Good to hear! Yeah, `run_iwyu_tests.py` picks up IWYU from the `PATH`, but you can pass the binary to use as an explicit argument like so:

    $ ./run_iwyu_tests.py -- ./build/bin/include-what-you-use

It's a little quirky, but I don't have any particularly bright ideas for perfection.

8 failures on Debian, btw? That's a little surprising, but let's deal with that separately.

vsapsai commented:
I still would like to try my approach from #109. But before I get to that, I'd like to point out that our mappings in some cases [override `#error` recommendation](https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_include_picker.cc#L406-L417).

kimgr commented:
Oh, OK. The built-in override is going to be a little tricky, since that happens before any `#error` directives are seen.

Let's close this PR and wait for your alternative patch. I think your approach has merit and solves more cases. I remember I was worried about textual includes/x-macros, but I think we can cross that bridge when we get there.

vsapsai commented:
Thanks for your patience. I was working last weekend so the patch isn't moving as fast as I'd like.
